keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -813,42 +866,23 @@ class SqlArtifactRepository(
         .from(ENVIRONMENT)
         .where(ENVIRONMENT.NAME.eq(environmentName))
         .and(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(deliveryConfig.uid))
-        .limit(1)
         .fetchOne(ENVIRONMENT.UID)
         ?: error("Environment '$environmentName not found")
 
       jooq
         .select(
           ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION,
           ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT,
-          ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS
+          ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS,
+          ENVIRONMENT_ARTIFACT_VERSIONS.REPLACED_BY,
+          ENVIRONMENT_ARTIFACT_VERSIONS.REPLACED_AT
         )
         .from(ENVIRONMENT_ARTIFACT_VERSIONS)
         .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
         .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifactUid))
         .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(version))
         .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.desc())
-        .limit(1)
-        .fetchOne { (version, deployedAt, promotionStatus) ->
-          val (replacedBy, replacedAt) = when (promotionStatus) {
-            CURRENT.name, PREVIOUS.name -> {
-              jooq
-                .select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION, ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT)
-                .from(ENVIRONMENT_ARTIFACT_VERSIONS)
-                .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
-                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifactUid))
-                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.ne(version))
-                .and(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.isNotNull)
-                .and(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.greaterThan(deployedAt))
-                .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.ne(VETOED.name))
-                .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.asc())
-                .limit(1)
-                .fetchOne { (replacedBy, replacedAt) ->
-                  Pair(replacedBy, replacedAt)
-                } ?: Pair(null, null)
-            }
-            else -> Pair(null, null)
-          }
+        .fetchOne { (version, deployedAt, promotionStatus, replacedBy, replacedAt) ->
```

Comment:
Oh! did I reinvent the wheel there?! Nooo... 😞

Reply:
Ah, no, I see it now. You're updating it in `markAsSuccessfullyDeployedTo`. Smart! 👍

Reply:
Yes! So we didn't have to do the nested query here! Just making an improvement :) plus it means that it's nicer to troubleshoot cause you can see things in the database